src/ksyms.cpp

```diff
@@ -23,8 +42,67 @@ std::string Ksyms::resolve(uint64_t addr, bool show_offset)
   return symbol.str();
 }
 
-Ksyms::~Ksyms()
+#ifdef USE_BLAZESYM
+std::optional<std::string> Ksyms::resolve_blazesym_int(uint64_t addr,
+                                                       bool show_offset)
 {
-  if (ksyms_)
-    bcc_free_symcache(ksyms_, -1);
+  if (symbolizer_ == nullptr) {
+    symbolizer_ = blaze_symbolizer_new();
+    if (symbolizer_ == nullptr)
+      return std::nullopt;
+  }
+
+  std::ostringstream symbol;
+  BLAZESYM_OPT blaze_symbolize_src_kernel src = {
+    .type_size = sizeof(src),
+    // Use default system-wide kallsyms file.
+    .kallsyms = NULL,
+    // Disable discovery and usage of kernel image.
+    // TODO: We should eventually support that, incorporating discovery logic
+    //       from find_vmlinux().
+    .kernel_image = "",
+  };
+  uint64_t addrs[1] = { addr };
+
+  const blaze_syms *syms = blaze_symbolize_kernel_abs_addrs(
+      symbolizer_, &src, addrs, ARRAY_SIZE(addrs));
+  if (syms == NULL)
+    return std::nullopt;
+
+  const blaze_sym *sym = &syms->syms[0];
+  if (sym->name == nullptr) {
+    blaze_syms_free(syms);
+    return std::nullopt;
+  }
+
+  symbol << sym->name;
+  if (show_offset) {
+    auto offset = addr - sym->addr;
+    symbol << "+" << offset;
+  }
+  blaze_syms_free(syms);
+  return symbol.str();
+}
+
+std::string Ksyms::resolve_blazesym(uint64_t addr, bool show_offset)
+{
+  if (auto sym = resolve_blazesym_int(addr, show_offset)) {
+    return *sym;
+  } else {
+    std::ostringstream symbol;
+    symbol << reinterpret_cast<void *>(addr);
+    return symbol.str();
```

Comment:
This is the same fallback as in `resolve_bcc`. Would it make sense to unify them (e.g. for the case we want to update it)?